tools/bitfield_gen.md

```diff
@@ -0,0 +1,653 @@
+}
+```
+
+This will generate a type `VMFault_t` in C. Assuming that the word size of the
+underlying architecture is 32 bits, the `VMFault_t` type above will have a size
+of two machine words. The sizes of the content and padding fields in a block
+must add up to a multiple of the machine word size.
+
+The bits in the block are laid out in memory in the order they are mentioned in
+the specification. That is, the left-most 32 bits (i.e. bits 63 to 32) of
+`VMFault` will correspond to the `address` field, the next 5 bits to `FSR`, the
+next 7 bits are inaccessible padding which is set to 0 on creation of the
+bitfield, then one bit for `instructionFault`, 15 bits of padding, and finally 4
+bits (bits 0 to 3 of the entire type) for `seL4_FaultType`.
+
+The following diagram corresponds to the block specification above.
+
+```bf_dia
+   │------------------------  address  ----------------------------│
+   ┌─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┐
+   │                                                               │ ..
+   └─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┘
+   63                                                             32
+
+                     instruction fault                  seL4_FaultType
+   │-- FSR --│             │-|                             │-------|
+   ┌─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┬─┐
+.. │         │             │ │                             │       │
+   └─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┴─┘
+    31      27              19                              3     0
+```
+
+For creating new values of the type `VMFault`, the generator will provide a
+function `VMFault` that takes an argument of the word size of the machine for
+the value of each field -- for the example above, this is `uint32_t`
+
+```c
+VMFault_t VMFault(uint32_t address, uint32_t FSR, uint32_t instructionFault,
+                  uint32_t seL4_FaultType);
+```
+
+Values that do not fit in the bit size of the target field will be cut off
+accordingly, i.e., as if using `val & mask(size)`.
+
+To read the value of each field, the generator will provide the following
+functions:
+
+```c
+uint32_t VMFault_get_address(VMFault_t vm_fault);
+uint32_t VMFault_get_FSR(VMFault_t vm_fault);
+uint32_t VMFault_get_instructionFault(VMFault_t vm_fault);
+uint32_t VMFault_get_seL4_FaultType(VMFault_t vm_fault);
+```
+
+For setting one of the fields in an existing value of type `VMFault_t`:
+
+```c
+VMFault_t VMFault_set_address(VMFault_t vm_fault, uint32_t address);
+VMFault_t VMFault_set_FSR(VMFault_t vm_fault, uint32_t FSR);
+VMFault_t VMFault_set_instructionFault(VMFault_t vm_fault, uint32_t instructionFault);
+VMFault_t VMFault_set_seL4_FaultType(VMFault_t vm_fault, uint32_t seL4_FaultType);
+```
+
+All of the above (constructor, getter, setter) are pure functions without side
+effects or dependency on global state. The generator declares all of them as
```

Comment:
Is this true? I think on some compiler configurations, there are `assert`s inserted that call abort if some arguments have invalid values...

Reply:
Yes, it does insert `asserts`, will mention those. They only have side effects when the code is incorrect ;-)